### PR TITLE
feat(streamer): added sponsored distribution support

### PR DESCRIPTION
**Pull request by keruch**
## Description

Part of https://github.com/dymensionxyz/dymension/issues/983

This PR adds x/sponsorship support for x/streamer:
* Added `sponsored` flag for the stream type
* Allowed creating sponsored streams
* Query x/sponsorship to get the distr for sponsored gauges at the end of the epoch

----

- [x]  Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [x]  Targeted PR against the correct branch
- [x]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x]  Linked to the GitHub issue with discussion and accepted design
- [x]  Targets only one GitHub issue
- [x]  Wrote unit and integration tests
- [x]  Wrote relevant migration scripts if necessary
- [x]  All CI checks have passed
- [x]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)
- [ ]  Updated the scripts for local run, e.g genesis_config_commands.sh if the PR changes parameters <-- in the next PR
- [x]  Add an issue in the [e2e-tests repo](https://github.com/dymensionxyz/e2e-tests) if necessary

SDK Checklist
- [x] Import/Export Genesis
- [x] Registered Invariants
- [x] Registered Events
- [x] Updated openapi.yaml
- [x] No usage of go `map`
- [x] No usage of `time.Now()`
- [x] Used fixed point arithmetic and not float arithmetic
- [x] Avoid panicking in Begin/End block as much as possible
- [x] No unexpected math Overflow
- [x] Used `sendCoin` and not `SendCoins`
- [x] Out-of-block compute is bounded
- [x] No serialized ID at the end of store keys
- [x] UInt to byte conversion should use BigEndian

Full security checklist [here](https://www.faulttolerant.xyz/2024-01-16-cosmos-security-1/)

----

For Reviewer:

- [ ]  Confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  Confirmed all author checklist items have been addressed

----

After reviewer approval:

- [ ]  In case the PR targets the main branch, PR should not be squash merge in order to keep meaningful git history.
- [ ]  In case the PR targets a release branch, PR must be rebased.

## Summary by CodeRabbit

- **New Features**
	- Introduced the ability to create sponsored streams within the application.
	- Added sponsorship-related parameters to stream creation proposals and distribution mechanisms.
	- Enhanced stream handling by supporting both sponsored and non-sponsored types.

- **Bug Fixes**
	- Improved testing for sponsored streams, ensuring accurate distribution calculations based on user votes.

- **Tests**
	- Expanded test coverage to include scenarios for both sponsored and non-sponsored streams.

### CODE REVIEW

File: x/streamer/types/distr_info.go [hunk 1]
```diff
@@ -32,3 +35,33 @@ func (r DistrRecord) ValidateBasic() error {
...
+	records := make([]DistrRecord, 0, len(d.Gauges))
+	for _, g := range d.Gauges {
+		// d.VotingPower is always > 0 to the Distribution type contract
+		weight := g.Power.Mul(hundred).Quo(d.VotingPower)
+
+		record := DistrRecord{
+			GaugeId: g.GaugeId,
+			Weight:  weight,
+		}
+		if err := record.ValidateBasic(); err != nil {
```
Comment (omritoptix) [inline x/streamer/types/distr_info.go:55]: if d.VotingPower > 0 seems like redundant check is redundant

File: x/streamer/types/distr_info.go [hunk 2]
```diff
@@ -32,3 +35,33 @@ func (r DistrRecord) ValidateBasic() error {
...
+			return nil, err
+		}
+
+		totalWeight = totalWeight.Add(weight)
```
Comment (omritoptix) [inline x/streamer/types/distr_info.go:59]: isn't totalWeight always 100 given sum(gagues.power) == VotingPower (which I assume is an invariant?)

File: x/streamer/keeper/keeper.go [hunk 3]
```diff
@@ -48,14 +50,27 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
...
+	var distrInfo *types.DistrInfo
+	if sponsored {
+		distr, err := k.sk.GetDistribution(ctx)
+		if err != nil {
+			return 0, fmt.Errorf("failed to get sponsorship distribution: %w", err)
+		}
+		distrInfo, err = types.DistrInfoFromDistribution(distr)
+		if err != nil {
+			return 0, fmt.Errorf("failed to compute distr info from sponsorship distribution: %w", err)
+		}
```
Comment (omritoptix) [inline x/streamer/keeper/keeper.go:67]: I guess can be dryed out with the code in `distribute`

File: CHANGELOG.md [hunk 4]
```diff
@@ -47,6 +47,8 @@ Ref: https://keepachangelog.com/en/1.0.0/
 - (rollapp) [#1036](https://github.com/dymensionxyz/dymension/issues/1036) Add VMType field to Rollapp
 - (swagger) [#856](https://github.com/dymensionxyz/dymension/issues/856) Add make command `proto-swagger-gen`
 - (delayedack) [#825](https://github.com/dymensionxyz/dymension/issues/825) Add query for rollapp packets using CLI
+- (sponsorship) [#983](https://github.com/dymensionxyz/dymension/issues/983) Add the sponsorship module to support community-driven distributions
```
Comment (omritoptix) [inline CHANGELOG.md:50]: changlog should be auto added by the changelog bot so no need for that

**Review by omritoptix — APPROVED** [review]

**Review by omritoptix — APPROVED** [review]

File: x/streamer/types/distr_info.go [hunk 5]
```diff
@@ -32,3 +35,30 @@ func (r DistrRecord) ValidateBasic() error {
 	}
 	return nil
 }
+
+var hundred = math.NewInt(100)
```
Comment (danwt) [inline x/streamer/types/distr_info.go:39]: dont do this it's weird

File: x/streamer/types/distr_info.go [hunk 6]
```diff
@@ -32,3 +35,30 @@ func (r DistrRecord) ValidateBasic() error {
...
+
+// DistrInfoFromDistribution converts sponsorship distribution to the DistrInfo type and performs a
+// basic validation for DistrInfo.Records. Returning an empty DistrInfo (with zero DistrInfo.TotalWeight)
+// is a valid scenario.
+func DistrInfoFromDistribution(d sponsorshiptypes.Distribution) *DistrInfo {
+	totalWeight := math.ZeroInt()
+	records := make([]DistrRecord, 0, len(d.Gauges))
+	for _, g := range d.Gauges {
+		// d.VotingPower is always > 0 to the Distribution type contract
+		weight := g.Power.Mul(hundred).Quo(d.VotingPower)
```
Comment (danwt) [inline x/streamer/types/distr_info.go:49]: percentage calculation? please indicate

**Review by danwt — APPROVED** [review]
Approved but could be tidied

**Pull request merged**